Add more rope related features

gta191977649 opened the pull request:
This PR attempt to implement some additional roped related feature in CObjects:

- `0796: GET_ROPE_HEIGHT_FOR_OBJECT` - [Object](https://library.sannybuilder.com/#/sa/classes/Object).GetRopeHeight(self: Object, var height: float)
- `0797: SET_ROPE_HEIGHT_FOR_OBJECT` - [Object](https://library.sannybuilder.com/#/sa/classes/Object).SetRopeHeight(self: Object, height: float)

working in progress

gta191977649 committed:
implmented GetRopeHeight

note: only works with object that has rope

lopezloo commented:
You might be interested:
https://github.com/lopezloo/mtasa-blue/tree/ropes
https://github.com/lopezloo/mtasa-blue/commits/ropes?author=lopezloo
https://github.com/lopezloo/mtasa-blue/blob/ropes/MTA10/game_sa/CRopeSA.h
https://github.com/lopezloo/mtasa-blue/blob/ropes/MTA10/game_sa/CRopeSA.cpp
https://github.com/lopezloo/mtasa-blue/blob/ropes/MTA10/game_sa/CRopesSA.h
https://github.com/lopezloo/mtasa-blue/blob/ropes/MTA10/game_sa/CRopesSA.cpp

I have implemented ropes in this branch. I'm not sure though if there is point in adding this to MTA since it can be implemented in Lua.

gta191977649 committed:
implemented SetRopeHeight

now SetRopeHeight should works
the reset is to figure out how dose the height exactly works, seems setting it to 1 will caused the magnet fall off.

gta191977649 commented:
now the `GetRopeHeight` and `SetRopeHeight` should work
the reset is to figure out how does the height exactly works, seems setting it to 1 will cause the magnet to fall off. 
NOTE: test it with object `1382`, other object id seems will crash the client, I haven't made the check in code yet.

gta191977649 commented:
> You might be interested: https://github.com/lopezloo/mtasa-blue/tree/ropes https://github.com/lopezloo/mtasa-blue/commits/ropes?author=lopezloo
> 
> I have implemented ropes in this branch. I'm not sure though if there is point in adding this to MTA since it can be implemented in Lua.



> You might be interested: https://github.com/lopezloo/mtasa-blue/tree/ropes https://github.com/lopezloo/mtasa-blue/commits/ropes?author=lopezloo
> 
> I have implemented ropes in this branch. I'm not sure though if there is point in adding this to MTA since it can be implemented in Lua.

maybe for someone who wants to re-create the crane, like in single-player, I know it can be done in pain Lua, but this is the vanllia sa methods :p

gta191977649 committed:
Merge branch 'master' into Rope-feature

tederis commented on Client/game_sa/CObjectSA.cpp:
```diff
@@ -401,6 +401,18 @@ void CObjectSA::SetScale(float fX, float fY, float fZ)
 
+float CObjectSA::GetRopeHeight()
+{
+    const auto ropeIndex = CRopesSA::FindRope(static_cast<CObjectSAInterface*>(this->GetInterface()));
```
What is the practical sense of explicitly use the **this->**?

tederis commented on Client/game_sa/CObjectSA.h:
```diff
@@ -155,6 +155,9 @@ class CObjectSA : public virtual CObject, public virtual CPhysicalSA
     CVector* GetScale();
     void     ResetScale();
 
+    float GetRopeHeight();
```
The method GetRopeHeight doesn't change anything in the class, right? The similar methods should be marked as **const**.

tederis commented on Client/game_sa/CRopesSA.cpp:
```diff
@@ -52,3 +52,20 @@ void CRopesSA::RemoveEntityRope(CEntitySAInterface* pEntity)
+    return -1;
+}
+
+CRopesSAInterface& CRopesSA::GetRope(int32 index)
```
What happens if the index becomes negative? All the function's paths must return a value. Otherwise we have an undefined behaviour, that's not a good thing.

tederis commented on Client/sdk/game/CObject.h:
```diff
@@ -41,4 +41,6 @@ class CObject : public virtual CPhysical
     virtual void     SetScale(float fX, float fY, float fZ) = 0;
     virtual CVector* GetScale() = 0;
     virtual void     ResetScale() = 0;
+    virtual float GetRopeHeight() = 0;
```
Similarly, the method should be const.

tederis commented on Client/mods/deathmatch/logic/CStaticFunctionDefinitions.cpp:
```diff
@@ -3900,6 +3900,22 @@ bool CStaticFunctionDefinitions::GetObjectElasticity(CClientObject& Object, floa
     return true;
 }
 
+bool CStaticFunctionDefinitions::GetRopeHeight(CClientObject& Object, float& height)
```
The first argument should be a const reference. Just because we want to get a value, but not change anything inside **Object**.
